compiler/rustc_hir_analysis/src/check/mod.rs

```diff
@@ -546,3 +554,80 @@ fn bad_non_zero_sized_fields<'tcx>(
 pub fn potentially_plural_count(count: usize, word: &str) -> String {
     format!("{} {}{}", count, word, pluralize!(count))
 }
+
+pub fn check_function_signature<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    cause_code: ObligationCauseCode<'tcx>,
+    fn_id: LocalDefId,
+    expected_sig: ty::PolyFnSig<'tcx>,
+) {
+    let param_env = ty::ParamEnv::empty();
+
+    let mut cause = ObligationCause::new(tcx.def_span(fn_id), fn_id, cause_code);
+
+    let infcx = &tcx.infer_ctxt().build();
+    let ocx = ObligationCtxt::new(infcx);
+
+    let unnormalized_actual_sig = infcx.instantiate_binder_with_fresh_vars(
+        cause.span,
+        infer::HigherRankedType,
+        tcx.fn_sig(fn_id).instantiate_identity(),
+    );
+
+    let norm_cause = ObligationCause::misc(cause.span, fn_id);
+    let actual_sig = ocx.normalize(&norm_cause, param_env, unnormalized_actual_sig);
+
+    let expected_sig = tcx.liberate_late_bound_regions(fn_id.into(), expected_sig);
+
+    match ocx.sup(&cause, param_env, expected_sig, actual_sig) {
```

Comment:
It concerns me that we're only checking the subtyping in one direction instead of both directions (with `eq` here).